BVTCK-114 Add tests for the support of group conversion for container elements + BVTCK-116

gsmet opened the pull request:
* https://hibernate.atlassian.net/browse/BVTCK-114
 * https://hibernate.atlassian.net/browse/BVTCK-116

Added to the TCK as it reused the same classes as for the original group conversion support.

I'll need to update the assertions once they are added to the BV spec.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/validation/groupconversion/RegisteredAddresses.java:
```diff
@@ -0,0 +1,106 @@
+
+	private final Map<Year, List<Address>> shipmentAddresses = new HashMap<>();
+
+	private final Map<Year, List<@Valid @ConvertGroup.List({
```
Are you using the explicit `@ConvertGroup.List` on purpose?

gsmet replied:
Not really. It came from another test and I let it this way hoping that it would avoid triggering the repeatable issue with Eclipse but it didn't work either. So we can probably remove them.

gsmet replied:
Removed.

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/tests/validation/groupconversion/RegisteredAddresses.java:
```diff
@@ -0,0 +1,106 @@
+
+	@Valid
+	@ConvertGroup(from = Default.class, to = BasicPostal.class)
+	public RegisteredAddresses(Map<Year, List<@Valid @ConvertGroup(from = Default.class, to = BasicPostal.class) Address>> mainAddresses) {
```
Hum, what are the interactions of these two `@Valid` actually? I think we need to clarify that in the spec. Filed BVAL-620 for this.

gsmet replied:
They are both taken into account.

As for now the global `ConvertGroup` is not applied to the container elements `@Valid`.

gunnarmorling replied:
Ok, so this validates each list element twice then, right? Wouldn't it be better for the sake of testing then to work with different conversion configurations?

gsmet replied:
Hmmm, I think we both misread the test. The first `@Valid`/`@ConvertGroup` is on the return value of the constructor, the other on the parameter. So they are not related at all.

gsmet replied:
Note that we should probably add a test for the case you have in mind but to be honest I'm still not sure how it should behave and we should probably discuss it next week when we have more time.

gunnarmorling replied:
Ah, yes, you are right. On the other, it'd be nice to have a test. I'll clarify this in the context of BVAL-620.

gsmet commented:
Force pushed an update to address the removal of the `ConvertGroup.List`

gunnarmorling commented on tests/src/main/java/org/hibernate/beanvalidation/tck/util/TestUtil.java:
```diff
@@ -167,16 +167,8 @@ public static void assertCorrectConstraintTypes(Set<? extends ConstraintViolatio
 		);
 
 		for ( String propertyPath : propertyPaths ) {
-			Path expectedPath = PathImpl.createPathFromString( propertyPath );
```
Hum, it seems risky to solely base the assertion on the string representation, given the spec doesn't mandate any particular one. Perhaps assertCorrectPropertyPaths() should take a more object-orientated representation of expected paths (we still can leave the current one for the most common case of plain non-indexed, non-keyed props). If you agree, can you open a follow-up BVTCK issue for it?

gsmet commented:
@gunnarmorling force pushed a new version addressing your comment about the property path string representation.

Note that for the next version, I'll incorporate assertj in the TCK and make the test utils consistent between HV and BVTCK. It will be easier to promote tests to the TCK and we'll have a better output in the assertions. It's too much of a pain now to have different behaviors.
